kolibri/core/discovery/utils/network/connections.py

```diff
@@ -0,0 +1,109 @@
+import socket
+from contextlib import closing
+
+from . import errors
+from .client import NetworkClient
+from .urls import parse_address_into_components
+from kolibri.core.utils.cache import CrossProcessCache
+
+
+class Nothing:
+    def __nonzero__(self):
+        return False
+
+    __bool__ = __nonzero__  # this is for python3
```

Comment:
I have some updates and tests for this to make sure these survive pickling, as they'll be stored in a cache.  Essentially the reason for this class is to make the connection caching logic more readable, as it's hard enough to follow already.   Basically it let's me store different kinds of falsey types in the cache as a way of keeping track different kinds of failures, but not having to make long, complicated boolean logic strings in the logic that interprets the cache.  Example:
```
INVALID_DEVICE_INFO = Nothing("invalid device info")
FAILED_TO_CONNECT = Nothing("failed to connect")

# these are different
assert INVALID_DEVICE_INFO != FAILED_TO_CONNECT

# but both are falsey, so you can stuff like this:
result = INVALID_DEVICE_INFO

if result:
    return result  # <---- this will never happen
elif result == FAILED_TO_CONNECT:
    print("failed to connect!")
   # ... this won't happen either
elif result == INVALID_DEVICE_INFO:
    print("invalid device info!")
    # now we can handle this specific kind of lack of a value correctly

```